README.md
## cloudwatch-metrics
This module provides a simplified wrapper for creating and publishing
CloudWatch metrics.

## Install
```
$ npm install cloudwatch-metrics
```
or
```
$ npm install cloudwatch-metrics --save
```

## Usage

### Initialization

By default, the library will log metrics to the `us-east-1` region and read
AWS credentials from the AWS SDK's [default environment variables](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html#Credentials_from_Environment_Variables).

If you want to change these values, you can call `initialize`:

```js
var cloudwatchMetrics = require('cloudwatch-metrics');
cloudwatchMetrics.initialize({
	region: 'us-east-1'
});
```

### Metric creation
For creating a metric, we simply need to provide the
namespace and the default type of metric:
```js
var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count');
```

### Metric creation - w/ default dimensions
If we want to add our own default dimensions, such as environment information,
we can add it in the following manner:
```js
var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
	Name: 'environment',
	Value: 'PROD'
}]);
```

### Metric creation - w/ options
If we want to disable a metric in certain environments (such as local development),
we can make the metric in the following manner:
```js
// isLocal is a boolean
var isLocal = someWayOfDetermingIfLocal();

var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
	Name: 'environment',
	Value: 'PROD'
}], {
	enabled: isLocal
});
```

The full list of configuration options is:

Option | Purpose
------ | -------
enabled | Whether or not we should send the metric to CloudWatch (useful for dev vs prod environments).
sendInterval | The interval in milliseconds at which we send any buffered metrics, defaults to 5000 milliseconds.
sendCallback | A callback to be called when we send metric data to CloudWatch (useful for logging any errors in sending data).
maxCapacity | A maximum number of events to buffer before we send immediately (before the sendInterval is reached).
withTimestamp | Include the timestamp with the metric value.
storageResolution | The metric storage resolution to use in seconds. Set to 1 for high resolution metrics. See (https://aws.amazon.com/blogs/aws/new-high-resolution-custom-metrics-and-alarms-for-amazon-cloudwatch/)

### Publishing metric data
Then, whenever we want to publish a metric, we simply do:
```js
myMetric.put(value, metric, units, additionalDimensions);
```

### Using summary metrics

Instead of sending individual data points for your metric, you may want to send
summary metrics. Summary metrics track statistics over time, and send those
statistics to CloudWatch on a configurable interval. For instance, you might
want to know your total network throughput, but you don't care about individual
request size percentiles. You could use `summaryPut` to track this data and send
it to CloudWatch with fewer requests:

```js
var metric = new cloudwatchMetrics.Metric('namespace', 'Bytes');

function onRequest(req) {
	// This will still track maximum, minimum, sum, count, and average, but won't
	// take up lots of CloudWatch requests doing so.
	metric.summaryPut(req.size, 'requestSize');
}
```

Note that metrics use different summaries for different dimensions, _and that
the order of the dimensions is significant!_ In other words, these track
different metric sets:

```js
var metric = new cloudwatchMetrics.Metric('namespace', 'Bytes');
// Different statistic sets!
metric.summaryPut(45, 'requestSize', [{Name: 'Region', Value: 'US'}, {Name: 'Server', Value: 'Card'}]);
metric.summaryPut(894, 'requestSize', [{Name: 'Server', Value: 'Card'}, {Name: 'Region', Value: 'US'}]);
```

### NOTES
Be aware that the `put` call does not actually send the metric to CloudWatch
at that moment. Instead, it stores unsent metrics and sends them to
CloudWatch on a predetermined interval (to help get around sending too many
metrics at once - CloudWatch limits you by default to 150 put-metric data
calls per second). The default interval is 5 seconds, if you want metrics
sent at a different interval, then provide that option when constructing your
CloudWatch Metric:

```js
var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
	Name: 'environment',
	Value: 'PROD'
}], {
	sendInterval: 3 * 1000 // It's specified in milliseconds.
});
```

You can also register a callback to be called when we actually send metrics
to CloudWatch - this can be useful for logging put-metric-data errors:
```js
var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
	Name: 'environment',
	Value: 'PROD'
}], {
	sendCallback: (err) => {
		if (!err) return;
		// Do your error handling here.
	}
});
```


CHANGELOG.md
CHANGELOG
=====
* 1.3.0 Allow setting unit on `put`, `putSummary` and `sample` #33 (thanks @Nevon!)
* 1.2.0 Add the ability to use summary metrics.
* 1.1.0 Add `metric.sample()`
* 1.0.0 Initial release.


test/index.js
var expect = require('chai').expect;
var _ = require('underscore');

var rewire = require('rewire');
var cloudwatchMetric = rewire('..');

describe('cloudwatch-metrics', function() {
  var restoreAWS;

  function attachHook(hook) {
    restoreAWS = cloudwatchMetric.__set__('AWS', {
      CloudWatch: function() {
        this.putMetricData = hook;
      }
    });
  }

  afterEach(function() {
    if (restoreAWS) {
      restoreAWS();
      restoreAWS = null;
    }
  });

  it('should buffer until timeout', function(done) {
    this.timeout(5000);
    attachHook(function(data, cb) {
      expect(data).to.deep.equal({
        MetricData: [{
          Dimensions: [{
            Name: "environment",
            Value: "PROD"
          }, {
            Name: "ExtraDimension",
            Value: "Value"
          }],
          MetricName: "metricName",
          Unit: "Count",
          Value: 1
        }],
        Namespace: 'namespace'
      });
      cb();
    });

    var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
      Name: 'environment',
      Value: 'PROD'
    }], {
      sendInterval: 1000,
      sendCallback: done
    });

    metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
  });

  it('should call continually', function(done) {
    this.timeout(3000);
    attachHook(function(data, cb) {
      expect(data).to.deep.equal({
        MetricData: [{
          Dimensions: [{
            Name: "environment",
            Value: "PROD"
          }, {
            Name: "ExtraDimension",
            Value: "Value"
          }],
          MetricName: "metricName",
          Unit: "Count",
          Value: 1
        }],
        Namespace: 'namespace'
      });
      cb();
    });

    var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
      Name: 'environment',
      Value: 'PROD'
    }], {
      sendInterval: 500,
      sendCallback: _.after(3, done)
    });

    var interval;
    var stop = _.after(3, () => clearInterval(interval));
    interval = setInterval(function() {
      metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
      stop();
    }, 400);
  });

  it('should buffer until the cap is hit', function(done) {
    attachHook(function(data, cb) {
      expect(data).to.deep.equal({
        MetricData: [{
          Dimensions: [{
            Name: "environment",
            Value: "PROD"
          }, {
            Name: "ExtraDimension",
            Value: "Value"
          }],
          MetricName: "metricName",
          Unit: "Count",
          Value: 1
        }, {
          Dimensions: [{
            Name: "environment",
            Value: "PROD"
          }, {
            Name: "ExtraDimension",
            Value: "Value"
          }],
          MetricName: "metricName",
          Unit: "Count",
          Value: 2
        }],
        Namespace: 'namespace'
      });
      cb();
    });

    var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
      Name: 'environment',
      Value: 'PROD'
    }], {
      sendInterval: 3000, // mocha defaults to a 2 second timeout so setting
                          // larger than that will cause the test to fail if we
                          // hit the timeout
      sendCallback: done,
      maxCapacity: 2
    });

    metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
    metric.put(2, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
  });
});


spec/src/indexSpec.js
/* globals describe, afterEach, it, expect, spyOn, jasmine */

var _ = require('underscore');

var rewire = require('rewire');
var cloudwatchMetric = rewire('../..');

describe('cloudwatch-metrics', function() {
  var restoreAWS;

  function attachHook(hook) {
    restoreAWS = cloudwatchMetric.__set__('CloudWatch', function() {
      this.putMetricData = hook;
    });
  }

  afterEach(function() {
    if (restoreAWS) {
      restoreAWS();
      restoreAWS = null;
    }
  });

  describe('put', function() {
    it('should buffer until timeout', function(done) {
      attachHook(function(data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            Value: 1
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        sendInterval: 1000,
        sendCallback: done
      });

      metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
    });

    it('should call continually', function(done) {
      attachHook(function(data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            Value: 1
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        sendInterval: 500,
        sendCallback: _.after(3, done)
      });

      var interval;
      var stop = _.after(3, () => clearInterval(interval));
      interval = setInterval(function() {
        metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
        stop();
      }, 400);
    });

    it('should buffer until the cap is hit', function(done) {
      attachHook(function(data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            Value: 1
          }, {
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            Value: 2
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        sendInterval: 3000, // mocha defaults to a 2 second timeout so setting
        // larger than that will cause the test to fail if we
        // hit the timeout
        sendCallback: done,
        maxCapacity: 2
      });

      metric.put(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
      metric.put(2, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}]);
    });

    it('should set a Timestamp if specified in the options', function(done) {
      attachHook(function(data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            Timestamp: jasmine.any(String),
            Value: 1
          }],
          Namespace: 'namespace'
        });
        expect(Date.parse(data.MetricData[0].Timestamp)).toBeLessThanOrEqual(Date.now());
        cb();
      });

      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        withTimestamp: true,
        sendInterval: 1000, // mocha defaults to a 2 second timeout so setting
        // larger than that will cause the test to fail if we
        // hit the timeout
        sendCallback: done,
      });

      metric.put(1, 'metricName', [{Name: 'ExtraDimension', Value: 'Value'}]);
    });

    it('should set a StorageResolution if specified in the options', function(done) {
      attachHook(function(data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Count',
            StorageResolution: 1,
            Value: 1
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        storageResolution: 1,
        sendInterval: 1000, // mocha defaults to a 2 second timeout so setting
        // larger than that will cause the test to fail if we
        // hit the timeout
        sendCallback: done,
      });

      metric.put(1, 'metricName', [{Name: 'ExtraDimension', Value: 'Value'}]);
    });

    it('should override the Unit from the namespace if specified in the put call', function (done) {
      attachHook(function (data, cb) {
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'metricName',
            Unit: 'Percent',
            Value: 1
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      const metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }], {
        sendInterval: 1000,
        sendCallback: done,
      });

      metric.put(1, 'metricName', 'Percent', [{ Name: 'ExtraDimension', Value: 'Value' }]);
    });
  });

  describe('sample', function() {
    it('should ignore metrics when not in the sample range', function() {
      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }]);

      spyOn(Math, 'random').and.returnValue(0.5);
      spyOn(metric, 'put');

      metric.sample(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}], 0.2);
      expect(metric.put).not.toHaveBeenCalled();
    });

    it('should call put when the we decide to sample a metric', function() {
      var metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD'
      }]);

      spyOn(Math, 'random').and.returnValue(0.1);
      // Just so we don't send anything to AWS.
      spyOn(metric, 'put').and.returnValue(undefined);

      metric.sample(1, 'metricName', [{Name:'ExtraDimension',Value: 'Value'}], 0.2);
      expect(metric.put).toHaveBeenCalled();
    });
  });

  describe('summaryPut', function() {
    it('should not call with no data', function(done) {
      attachHook(() => {
        throw new Error('should not get send callback');
      });

      const metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD',
      }], {
        summaryInterval: 100,
        sendCallback() {
          throw new Error('should not get send callback');
        },
      });

      spyOn(metric, '_summarizeMetrics');

      setTimeout(() => {
        expect(metric._summarizeMetrics).toHaveBeenCalled();
        done();
      }, 250);
    });

    it('should call with summary', function(done) {
      let hookCalls = 0;
      attachHook((data, cb) => {
        ++hookCalls;
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'some-metric',
            Unit: 'Count',
            StatisticValues: {
              Minimum: 12,
              Maximum: 13,
              Sum: 25,
              SampleCount: 2,
            },
          }, {
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'some-other-metric',
            Unit: 'Count',
            StatisticValues: {
              Minimum: 2,
              Maximum: 2,
              Sum: 2,
              SampleCount: 1,
            },
          }, {
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'a-metric-with-different-unit',
            Unit: 'Percent',
            StatisticValues: {
              Minimum: 5,
              Maximum: 5,
              Sum: 5,
              SampleCount: 1,
            },
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      const metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD',
      }], {
        summaryInterval: 100,
        sendCallback() {
          expect(hookCalls).toBe(1);
          done();
        },
      });

      metric.summaryPut(12, 'some-metric', [{ Name: 'ExtraDimension', Value: 'Value' }]);
      metric.summaryPut(2, 'some-other-metric', [{ Name: 'ExtraDimension', Value: 'Value' }]);
      metric.summaryPut(5, 'a-metric-with-different-unit', 'Percent', [{ Name: 'ExtraDimension', Value: 'Value' }]);
      setTimeout(() => {
        metric.summaryPut(13, 'some-metric', [{Name: 'ExtraDimension', Value: 'Value'}]);
      }, 50);
    });

    it('should call after no data', function(done) {
      let hookCalls = 0;
      attachHook((data, cb) => {
        ++hookCalls;
        expect(data).toEqual({
          MetricData: [{
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'some-metric',
            Unit: 'Count',
            StatisticValues: {
              Minimum: 12,
              Maximum: 13,
              Sum: 25,
              SampleCount: 2,
            },
          }, {
            Dimensions: [{
              Name: 'environment',
              Value: 'PROD'
            }, {
              Name: 'ExtraDimension',
              Value: 'Value'
            }],
            MetricName: 'some-other-metric',
            Unit: 'Count',
            StatisticValues: {
              Minimum: 2,
              Maximum: 2,
              Sum: 2,
              SampleCount: 1,
            },
          }],
          Namespace: 'namespace'
        });
        cb();
      });

      const metric = new cloudwatchMetric.Metric('namespace', 'Count', [{
        Name: 'environment',
        Value: 'PROD',
      }], {
        summaryInterval: 200,
        sendCallback() {
          expect(hookCalls).toBe(1);
          done();
        },
      });

      setTimeout(() => {
        metric.summaryPut(12, 'some-metric', [{Name: 'ExtraDimension', Value: 'Value'}]);
        metric.summaryPut(2, 'some-other-metric', [{Name: 'ExtraDimension', Value: 'Value'}]);
        setTimeout(() => {
          metric.summaryPut(13, 'some-metric', [{Name: 'ExtraDimension', Value: 'Value'}]);
        }, 50);
      }, 300);
    });
  });
});


src/summarySet.js
/**
 * A simple tool to track the aggregate of an unordered sequence of values.
 *
 * Produces MIN, MAX, SUM, and COUNT.
 */
class SummarySet {
  constructor({minSentinel = Infinity, maxSentinel = -Infinity} = {}) {
    this._minSentinel = minSentinel;
    this._maxSentinel = maxSentinel;

    this.reset();
  }

  put(value) {
    this._sum += value;
    this._min = Math.min(this._min, value);
    this._max = Math.max(this._max, value);
    ++this._count;
  }

  get size() {
    return this._count;
  }

  /**
   * Get and reset the summarized statistics.
   *
   * @return {Object<String, Number>} The Minimum, Maximum, Sum, and SampleCount values.
   */
  get() {
    const result = this.peek();
    this.reset();
    return result;
  }

  /**
   * Get the summarized statistics, but do not reset them.
   *
   * @return {Object<String, Number>} The Minimum, Maximum, Sum, and SampleCount values.
   */
  peek() {
    return {
      Minimum: this._min,
      Maximum: this._max,
      Sum: this._sum,
      SampleCount: this._count,
    };
  }

  /**
   * Reset the summarized statistics.
   */
  reset() {
    this._min = this._minSentinel;
    this._max = this._maxSentinel;
    this._sum = 0;
    this._count = 0;
  }
}

module.exports = SummarySet;


index.js
/**
 * This module provides a simplified wrapper for creating and publishing
 * CloudWatch metrics. We should always initialize our environment first:
 *
 * ```
 * var cloudwatchMetrics = require('cloudwatch-metrics');
 * cloudwatchMetrics.initialize({
 * 	region: 'us-east-1'
 * });
 * ```
 *
 * For creating a metric, we simply need to provide the
 * namespace and the type of metric:
 *
 * ```
 * 	var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count');
 * ```
 *
 * If we want to add our own default dimensions, such as environment information,
 * we can add it in the following manner:
 *
 * ```
 * var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
 * 	Name: 'environment',
 * 	Value: 'PROD'
 * }]);
 * ```
 *
 * If we want to disable a metric in certain environments (such as local development),
 * we can make the metric in the following manner:
 *
 * ```
 * // isLocal is a boolean
 * var isLocal = someWayOfDetermingIfLocal();
 *
 * var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
 * 	Name: 'environment',
 * 	Value: 'PROD'
 * }], {
 * 	enabled: isLocal
 * });
 * ```
 *
 * Then, whenever we want to publish a metric, we simply do:
 *
 * ```
 * myMetric.put(value, metric, additionalDimensions);
 * ```
 *
 * Be aware that the `put` call does not actually send the metric to CloudWatch
 * at that moment. Instead, it stores unsent metrics and sends them to
 * CloudWatch on a predetermined interval (to help get around sending too many
 * metrics at once - CloudWatch limits you by default to 150 put-metric data
 * calls per second). The default interval is 5 seconds, if you want metrics
 * sent at a different interval, then provide that option when construction your
 * CloudWatch Metric:
 *
 * ```
 * var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
 * 	Name: 'environment',
 * 	Value: 'PROD'
 * }], {
 * 	sendInterval: 3 * 1000 // It's specified in milliseconds.
 * });
 * ```
 *
 * You can also register a callback to be called when we actually send metrics
 * to CloudWatch - this can be useful for logging put-metric-data errors:
 * ```
 * var myMetric = new cloudwatchMetrics.Metric('namespace', 'Count', [{
 * 	Name: 'environment',
 * 	Value: 'PROD'
 * }], {
 * 	sendCallback: (err) => {
 * 		if (!err) return;
 * 		// Do your error handling here.
 * 	}
 * });
 * ```
 */

var CloudWatch = require('aws-sdk/clients/cloudwatch');
const SummarySet = require('./src/summarySet');

var _awsConfig = {region: 'us-east-1'};
/**
 * setIndividialConfig sets the default configuration to use when creating AWS
 * metrics. It defaults to simply setting the AWS region to `us-east-1`, i.e.:
 *
 * {
 * 	region: 'us-east-1'
 * }
 * @param {Object} config The AWS SDK configuration options one would like to set.
 */
function initialize(config) {
  _awsConfig = config;
}

const DEFAULT_METRIC_OPTIONS = {
  enabled: true,
  sendInterval: 5000,
  summaryInterval: 10000,
  sendCallback: () => {},
  maxCapacity: 20,
  withTimestamp: false,
  storageResolution: undefined
};

/**
 * Create a custom CloudWatch Metric object that sets pre-configured dimensions and allows for
 * customized metricName and units. Each CloudWatchMetric object has it's own internal
 * AWS.CloudWatch object to prevent errors due to overlapping callings to
 * AWS.CloudWatch#putMetricData.
 *
 * @param {String} namespace         CloudWatch namespace
 * @param {String} units             CloudWatch units
 * @param {Object} defaultDimensions (optional) Any default dimensions we'd
 *    like the metric to have.
 * @param {Object} options           (optional) Options used to control metric
 *    behavior.
 *   @param {Bool} options.enabled   Defaults to true, controls whether we
 *      publish the metric when `Metric#put()` is called - this is useful for
 *      turning off metrics in specific environments.
 */
function Metric(namespace, units, defaultDimensions, options) {
  var self = this;
  self.cloudwatch = new CloudWatch(_awsConfig);
  self.namespace = namespace;
  self.units = units;
  self.defaultDimensions = defaultDimensions || [];
  self.options = Object.assign({}, DEFAULT_METRIC_OPTIONS, options);
  self._storedMetrics = [];
  this._summaryData = new Map();

  if (self.options.enabled) {
    self._interval = setInterval(() => {
      self._sendMetrics();
    }, self.options.sendInterval);

    this._summaryInterval = setInterval(() => {
      this._summarizeMetrics();
    }, this.options.summaryInterval);
  }
}

/**
 * Publish this data to Cloudwatch
 * @param {Integer|Long} value          Data point to submit
 * @param {String} namespace            Name of the metric
 * @param {String} units                CloudWatch units
 * @param {Array} [additionalDimensions]  Array of additional CloudWatch metric dimensions. See
 * http://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_Dimension.html for details.
 */
Metric.prototype.put = function(...args) {
  if (args.length === 3) {
    const [value, metricName] = args;
    const shouldInheritUnits = Array.isArray(args[2]);
    const units = shouldInheritUnits ? this.units : args[2];
    const additionalDimensions = shouldInheritUnits ? args[2] : [];
    return this._put(value, metricName, units, additionalDimensions);
  } else if (args.length === 2) {
    return this._put(...args, this.units);
  }
  return this._put(...args);
};

Metric.prototype._put = function(value, metricName, units, additionalDimensions) {
  var self = this;
  // Only publish if we are enabled
  if (self.options.enabled) {
    var payload = {
      MetricName: metricName,
      Dimensions: self.defaultDimensions.concat(additionalDimensions),
      Unit: units,
      Value: value
    };
    if (this.options.withTimestamp) {
      payload.Timestamp = new Date().toISOString();
    }
    if (this.options.storageResolution) {
      payload.StorageResolution = this.options.storageResolution;
    }

    self._storedMetrics.push(payload);

    // We need to see if we're at our maxCapacity, if we are - then send the
    // metrics now.
    if (self._storedMetrics.length === self.options.maxCapacity) {
      clearInterval(self._interval);
      self._sendMetrics();
      self._interval = setInterval(() => {
        self._sendMetrics();
      }, self.options.sendInterval);
    }
  }
};

/**
 * Summarize the data using a statistic set and put it on the configured summary interval. This will
 * cause Cloudwatch to be unable to track the value distribution, so it'll only show summation and
 * bounds. The order of additionalDimensions is important, and rearranging the order will cause the
 * Metric instance to track those two summary sets independently!
 * @param {Number} value The value to include in the summary.
 * @param {String} metricName The name of the metric we're summarizing.
 * @param {String} units CloudWatch units
 * @param {Object[]} additionalDimensions The extra dimensions we're tracking.
 */
Metric.prototype.summaryPut = function(...args) {
  if (args.length === 3) {
    const [value, metricName] = args;
    const shouldInheritUnits = Array.isArray(args[2]);
    const units = shouldInheritUnits ? this.units : args[2];
    const additionalDimensions = shouldInheritUnits ? args[2] : [];
    return this._summaryPut(value, metricName, units, additionalDimensions);
  } else if (args.length === 2) {
    return this._summaryPut(...args, this.units, []);
  }
  return this._summaryPut(...args);
};

Metric.prototype._summaryPut = function(value, metricName, units, additionalDimensions = []) {
  const key = makeKey(metricName, units, additionalDimensions);
  const entry = this._summaryData.get(key);

  let set;
  if (entry) {
    set = entry[3];
  } else {
    set = new SummarySet();
    const allDimensions = [...this.defaultDimensions, ...additionalDimensions];
    this._summaryData.set(key, [metricName, units, allDimensions, set]);
  }
  set.put(value);
};

/**
 * Samples a metric so that we send the metric to Cloudwatch at the given
 * sampleRate.
 * @param {Integer|Long} value          Data point to submit
 * @param {String} namespace            Name of the metric
 * @param {String} units                CloudWatch units
 * @param {Array} additionalDimensions  Array of additional CloudWatch metric dimensions. See
 * http://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_Dimension.html for details.
 * @param  {Float} sampleRate           The rate at which to sample the metric at.
 *    The sample rate must be between 0.0 an 1.0. As an example, if you provide
 *    a sampleRate of 0.1, then we will send the metric to Cloudwatch 10% of the
 *    time.
 */
Metric.prototype.sample = function(...args) {
  if (args.length === 4) {
    const [value, metricName, additionalDimensions, sampleRate] = args;
    const units = this.units;
    return this._sample(value, metricName, units, additionalDimensions, sampleRate);
  }

  return this.prototype._sample(...args);
};

Metric.prototype._sample = function(value, metricName, units, additionalDimensions, sampleRate) {
  sampleRate = Array.isArray(additionalDimensions) ? sampleRate : additionalDimensions;
  if (Math.random() < sampleRate) this.put(value, metricName, units, additionalDimensions);
};

/**
 * _sendMetrics is called on a specified interval (defaults to 5 seconds but
 * can be overridden but providing a `sendInterval` option when creating a
 * Metric). It is what actually sends metrics to CloudWatch. It passes the
 * sendCallback option (if provided) as the callback to the put-metric-data
 * call. This can be useful for logging AWS errors.
 */
Metric.prototype._sendMetrics = function() {
  var self = this;
  // NOTE: this would be racy except that NodeJS is single threaded.
  const dataPoints = self._storedMetrics;
  self._storedMetrics = [];

  if (!dataPoints || !dataPoints.length) return;

  self.cloudwatch.putMetricData({
    MetricData: dataPoints,
    Namespace: self.namespace
  }, self.options.sendCallback);
};

/**
 * Shuts down metric service by clearing any outstanding timer and sending any existing metrics
 */
Metric.prototype.shutdown = function() {
  clearInterval(this._interval);
  this._sendMetrics();
};

/**
 * Gets whether outstanding metrics exist or not.
 *
 * @return {boolean}
 */
Metric.prototype.hasMetrics = function() {
  return !!this._storedMetrics.length;
};

/**
 * _summarizeMetrics is called on a specified interval (default, 10 seconds). It
 * sends summarized statistics to Cloudwatch.
 */
Metric.prototype._summarizeMetrics = function() {
  const summaryEntries = this._summaryData.values();
  const dataPoints = [];
  for (const [MetricName, Unit, Dimensions, set] of summaryEntries) {
    if (!set.size) continue;

    dataPoints.push({
      MetricName,
      Dimensions,
      StatisticValues: set.get(),
      Unit,
    });

    if (dataPoints.length === this.options.maxCapacity) {
      // Put a copy of the points we've gathered, then empty the array so we can
      // get more.
      this._putSummaryMetrics(dataPoints.slice());
      dataPoints.length = 0;
    }
  }

  if (dataPoints.length) {
    this._putSummaryMetrics(dataPoints);
  }
};

/**
 * Put a single batch of summarized metrics to Cloudwatch. This helps avoid
 * hitting the Cloudwatch per-call maximum.
 */
Metric.prototype._putSummaryMetrics = function(MetricData) {
  this.cloudwatch.putMetricData({
    MetricData,
    Namespace: this.namespace,
  }, this.options.sendCallback);
};

/**
 * Make a key for a given metric name and some dimensions. This works on the
 * assumption that sane people won't put a null character in their metric name
 * or dimension name/value.
 *
 * @param {String} metricName
 * @param {String} units
 * @param {Object[]} dimensions
 * @returns {String} Something we can actually use as a Map key.
 */
function makeKey(metricName, units, dimensions) {
  let key = `${metricName}\0${units}`;
  for (const {Name, Value} of dimensions) {
    key += `\0${Name}\0${Value}`;
  }
  return key;
}

module.exports = {
  initialize,
  Metric
};
